esmvaltool/preprocessor/_multimodel.py

```diff
@@ -13,42 +26,330 @@
     bounds = [range(int(b[0]), int(b[-1]) + 1) for b in all_times]
-    time_pts = reduce(np.intersect1d, (i for i in bounds))
+    time_pts = reduce(np.intersect1d, bounds)
     if len(time_pts) > 1:
-        return _sdat(float(time_pts[0])), _sdat(float(time_pts[-1]))
+        time_bounds_list = [time_pts[0], time_pts[-1]]
+        return time_bounds_list
 
 
 def _slice_cube(cube, min_t, max_t):
     """slice cube on time"""
     fmt = '%Y-%m-%d-%H'
     ctr = iris.Constraint(time=lambda x:
-                          min_t <= dd.strptime(x.point.strftime(fmt),
-                                               fmt) <= max_t)
+                          min_t <=
+                          datetime.strptime(x.point.strftime(fmt),
+                                            fmt) <= max_t)
     cube_slice = cube.extract(ctr)
     return cube_slice
 
 
-def multi_model_mean(cubes, span, filename, exclude):
+def _slice_cube2(cube, t_1, t_2):
+    """
+    Efficient slicer
+
+    Slice cube on time more memory-efficiently than
+    _slice_cube(); using this function adds virtually
+    no extra memory to the process
+    """
+    time_pts = [t for t in cube.coord('time').points]
+    converted_t = _time_pts_by_calendar(cube)
+    idxs = sorted([time_pts.index(ii)
+                   for ii, jj in zip(time_pts, converted_t)
+                   if t_1 <= jj <= t_2])
+    cube_t_slice = cube.data[idxs[0]:idxs[-1] + 1]
+    return cube_t_slice
+
+
+def _monthly_t(cubes):
+    """Rearrange time points for monthly data"""
+    # get original cubes tpoints
+    tpts = []
+    # make all 365-day calendards
+    for cube in cubes:
+        tpts.append([int(t) for t in _time_pts_by_calendar(cube)])
+
+    # convert to months for MONTHLY data
+    t_x = list(set().union(*tpts))
+    t_x = list({int((a / 365.) * 12.) for a in t_x})
+    t_x.sort()
+    # remake the time axis for roughly the 15th of the month
+    t_x0 = list({t * 365. / 12. + 15. for t in t_x})
+    t_x0.sort()
+
+    return t_x, t_x0
+
+
+def _full_time(cubes):
+    """Construct a contiguous collection over time"""
+    datas = []
+    # get rearranged time points
+    t_x, t_x0 = _monthly_t(cubes)
+    # loop through cubes and apply masks
+    for cube in cubes:
+        # recast time points
+        time_redone = [int(t) for t in _time_pts_by_calendar(cube)]
+        # construct new shape
+        fine_shape = tuple([len(t_x)] + list(cube.data.shape[1:]))
+        # find indices of present time points
+        oidx = [
+            t_x.index(int((s / 365.) * 12.))
+            for s in time_redone]
+        # reshape data to include all possible times
+        ndat = np.ma.resize(cube.data, fine_shape)
+        # build the time mask
+        c_ones = np.ones(fine_shape, bool)
+        c_ones[oidx] = False
+        ndat.mask |= c_ones
+
+        # stitch new datas
+        datas.append(ndat)
+
+    # return datas
+    return datas, t_x0
+
+
+def _apply_overlap(cube, tx1, tx2):
```

Comment:
Is this function still needed?

Reply:
well, same story here as for _slice_cube(), it's done only once for the template cube